src/backend/opencl/kernel/mean.hpp

```diff
@@ -0,0 +1,698 @@
+            options << " -D USE_DOUBLE";
+        }
+
+        const char *ker_strs[] = {mops_cl, mean_first_cl};
+        const int   ker_lens[] = {mops_cl_len, mean_first_cl_len};
+        Program prog;
+        buildProgram(prog, 2, ker_strs, ker_lens, options.str());
+        entry.prog = new Program(prog);
+        entry.ker = new Kernel(*entry.prog, "mean_first_kernel");
+
+        addKernelToCache(device, ref_name, entry);
+    }
+
+    NDRange local(threads_x, THREADS_PER_GROUP / threads_x);
+    NDRange global(groups_x * in.info.dims[2] * local[0],
+            groups_y * in.info.dims[3] * local[1]);
+
+    uint repeat = divup(in.info.dims[0], (local[0] * groups_x));
+
+    if (input_weight && output_weight) {
+        auto meanOp = KernelFunctor<
+            Buffer, KParam,
+            Buffer, KParam,
+            Buffer, KParam,
+            Buffer, KParam,
+            uint, uint, uint>(*entry.ker);
+        meanOp(EnqueueArgs(getQueue(), global, local),
+                *out.data, out.info,
+                *owt.data, owt.info,
+                *in.data, in.info,
+                *iwt.data, iwt.info,
+                groups_x, groups_y, repeat);
+    } else if (!input_weight && !output_weight) {
+        auto meanOp = KernelFunctor<
+            Buffer, KParam,
+            Buffer, KParam,
+            uint, uint, uint>(*entry.ker);
+        meanOp(EnqueueArgs(getQueue(), global, local),
+                *out.data, out.info,
+                *in.data, in.info,
+                groups_x, groups_y, repeat);
+    } else if ( input_weight && !output_weight) {
+        auto meanOp = KernelFunctor<
+            Buffer, KParam,
+            Buffer, KParam,
+            Buffer, KParam,
+            uint, uint, uint>(*entry.ker);
+        meanOp(EnqueueArgs(getQueue(), global, local),
+                *out.data, out.info,
+                *in.data, in.info,
+                *iwt.data, iwt.info,
+                groups_x, groups_y, repeat);
+    } else if (!input_weight &&  output_weight) {
+        auto meanOp = KernelFunctor<
+            Buffer, KParam,
+            Buffer, KParam,
+            Buffer, KParam,
+            uint, uint, uint>(*entry.ker);
+        meanOp(EnqueueArgs(getQueue(), global, local),
+                *out.data, out.info,
+                *owt.data, owt.info,
+                *in.data, in.info,
+                groups_x, groups_y, repeat);
+    }
+
+    CL_DEBUG_FINISH(getQueue());
+}
+
+template<typename Ti, typename Tw, typename To>
+void mean_first(Param out, Param in, Param iwt)
+{
+    uint threads_x = nextpow2(std::max(32u, (uint)in.info.dims[0]));
+    threads_x = std::min(threads_x, THREADS_PER_GROUP);
+    uint threads_y = THREADS_PER_GROUP / threads_x;
+
+    uint groups_x = divup(in.info.dims[0], threads_x * REPEAT);
+    uint groups_y = divup(in.info.dims[1], threads_y);
+
+    Param tmpOut = out;
+    Param tmpWt;
+    tmpWt.info.offset = 0;
+    for (int k = 0; k < 4; ++k) {
```

Comment:
This is just creating an Array<T> object. Use Array<T> instead of creating params. Array<T> has implicit conversions to Param objects.